java/src/main/native/src/row_conversion.cu

```diff
@@ -0,0 +1,636 @@
+                shared_index < shared_input_end;
+                shared_index += shared_input_stride) {
+            long_output[start_output_index + shared_index] = long_shared[shared_index];
+        }
+        __syncthreads();
+        // Go for the next round
+    }
+}
+
+/**
+ * Calculate the dimensions of the kernel for fixed width only columns.
+ * @param [in] num_columns the number of columns being copied.
+ * @param [in] num_rows the number of rows being copied.
+ * @param [in] size_per_row the size each row takes up when padded.
+ * @param [out] blocks the size of the blocks for the kernel
+ * @param [out] threads the size of the threads for the kernel
+ * @return the size in bytes of shared memory needed for each block.
+ */
+static int calc_fixed_width_kernel_dims(
+    const cudf::size_type num_columns,
+    const cudf::size_type num_rows,
+    const cudf::size_type size_per_row,
+    dim3 & blocks,
+    dim3 & threads) {
+
+    // We have found speed degrades when a thread handles more than 4 columns.
+    // Each block is 2 dimensional. The y dimension indicates the columns.
+    // We limit this to 32 threads in the y dimension so we can still
+    // have at least 32 threads in the x dimension (1 warp) which should
+    // result in better coalescing of memory operations. We also
+    // want to guarantee that we are processing a multiple of 32 threads
+    // in the x dimension because we use atomic operations at the block
+    // level when writing validity data out to main memory, and that would
+    // need to change if we split a word of validity data between blocks.
+    int y_block_size = (num_columns + 3) / 4;
+    if (y_block_size > 32) {
+      y_block_size = 32;
+    }
+    int x_possible_block_size = 1024/y_block_size;
+    // 48KB is the default setting for shared memory per block according to the cuda tutorials
+    // If someone configures the GPU to only have 16 KB this might not work.
+    int max_shared_size = 48 * 1024;
+    int max_block_size = max_shared_size/size_per_row;
+    // If we don't have enough shared memory there is no point in having more threads
+    // per block that will just sit idle
+    max_block_size = max_block_size > x_possible_block_size ? x_possible_block_size : max_block_size;
+    // Make sure that the x dimension is a multiple of 32 this not only helps
+    // coalesce memory access it also lets us do a ballot sync for validity to write
+    // the data back out the warp level.  If x is a multiple of 32 then each thread in the y
+    // dimension is associated with one or more warps, that should correspond to the validity
+    // words directly.
+    int block_size = (max_block_size / 32) * 32;
+    CUDF_EXPECTS(block_size != 0, "Row size is too large to fit in shared memory");
+
+    int num_blocks = (num_rows + block_size - 1) / block_size;
+    if (num_blocks < 1) {
+        num_blocks = 1;
+    } else if (num_blocks > 10240) {
+        num_blocks = 10240;
```

Comment:
This function is informative.

Why is there a cap (`10240`) for the number of blocks in the grid?

Reply:
I picked an arbitrary magic number.  I should probably try to remove it because it does not matter.  When I first read about blocks it said that CUDA had a limit of 65,535 blocks in the x dimension, and I thought that hitting a high number like that probably added a lot of overhead, but it turns out that we require compute capability 6.0 or above which actually has a limit of `2^31 - 1` for the x dimension in blocks, so I should play around with it a bit more.  That is essentially the same as the number of rows that we can support in a single column.  I'll play around with it and see if I run into any issues if I let it go very large.